src/sentry/utils/sentryappwebhookrequests.py

```diff
@@ -0,0 +1,114 @@
+from __future__ import absolute_import
+
+import six
+import json
+
+from dateutil.parser import parse as parse_date
+from django.conf import settings
+from django.utils import timezone
+
+from sentry.utils import redis
+from sentry.models.sentryapp import VALID_EVENTS
+from sentry.models import Organization
+
+BUFFER_SIZE = 100
+
+
+class SentryAppWebhookRequestsBuffer(object):
+    """
+    Create a data structure to store basic information about Sentry App webhook requests in Redis
+    This should store the last 100 requests and last 100 errors (in different keys) for each event type, for each Sentry App
+    """
+
+    def __init__(self, sentry_app):
+        self.sentry_app = sentry_app
+
+        cluster_id = getattr(settings, "SENTRY_WEBHOOK_LOG_REDIS_CLUSTER", None)
+        if cluster_id is None:
+            self.client = redis.clusters.get("default").get_local_client(0)
+        else:
+            self.client = redis.redis_clusters.get(cluster_id)
```

Comment:
@tkaemming when testing on my local machine this wouldn't work with redis_clusters even with a default value. I'm not sure if I need to add something to the config file in getsentry or sentry for this to work...

Reply:
Ok, so — there are three other places that we use Redis Cluster: attachments cache, similarity and quotas.

Similarity is a non-essential feature (it's flagged off by default for development and on-premise installations), and fails over to a dummy backend if the provided backend isn't a Redis Cluster client: https://github.com/getsentry/sentry/blob/aff783bb182e98cef3068dfd4801ff7f1005b864/src/sentry/similarity/__init__.py#L66-L71

The attachment cache and quotas both accept *either* a rb or Redis Cluster client:

https://github.com/getsentry/sentry/blob/aff783bb182e98cef3068dfd4801ff7f1005b864/src/sentry/attachments/redis.py#L13-L27

https://github.com/getsentry/sentry/blob/17bf005b1c51ea20e617ed9c273c68eabd718260/src/sentry/quotas/redis.py#L132-L141

For quotas, this was done as a transitional move for us in production, and I think we basically consider that usage of rb deprecated at this point (though we have not established any specific timeline or strategy for forcing that upgrade in development with devservices, or on-premise usage.) I'm less sure about attachments, but my guess is that is/was probably a similar situation.

In general, I think that we are pretty hesitant to introduce a new dependency on rb, since that adds just one more thing to migrate in the future (not just for us, but also for on-premise uses.) With that in mind, is this something that can be optional for on-premise use if no Redis Cluster is configured and depend exclusively on Redis Cluster? 

- If yes: Great, let's remove the `redis.clusters` bits, and make sure whatever is calling this can deal with it not being available for use. Easy.
- If not: Okay, but this does mean we're incurring debt here that we'll have to collect later. Also, there's also no guarantee that `default` is either a rb or Redis Cluster instance. This will need to look up the value of `getattr(settings, "SENTRY_WEBHOOK_LOG_REDIS_CLUSTER", None)` in both `redis.clusters` and `redis.redis_clusters` and use whichever one exists.

In any case, we probably need to start working with @BYK for on-premise and @mattrobenolt for single tenant about figuring out how we will force the eventual migration to Redis Cluster.

Reply:
@tkaemming Talked with @scefali and decided that we can just have this optional for on-premise users. In this comment, are you saying that it could potentially work for on-premise users if they do have Redis Cluster configured though?

And once we remove the `redis.clusters` stuff, do I still need to be adding some sort of configuration somewhere (like I don't know where the `SENTRY_WEBHOOK_LOG_REDIS_CLUSTER` is coming from because I just made up that string and didn't put it anywhere else). 

Also does local development/the devserver not support Redis clusters at all? Or is there some way to test it?

Reply:
I guess my comment didn't go here like I had intended, so please refer to this: https://github.com/getsentry/sentry/pull/15269#issuecomment-548077147

Reply:
> In this comment, are you saying that it could potentially work for on-premise users if they do have Redis Cluster configured though?

Yes. There'd be nothing preventing this from working on-premise as long as they have the correct type of Redis backend.

> And once we remove the `redis.clusters` stuff, do I still need to be adding some sort of configuration somewhere (like I don't know where the `SENTRY_WEBHOOK_LOG_REDIS_CLUSTER` is coming from because I just made up that string and didn't put it anywhere else).

Yes. That setting name is fine. I'd continue to use `default` as the default, but only check if it exists in `redis.redis_clusters`. This may throw an exception if the `default` cluster is an rb cluster, so something will need to catch that exception and fall back to the path where this is disabled if that happens.

> Also does local development/the devserver not support Redis clusters at all? Or is there some way to test it?

It does, but not by default. One would need to be explicitly configured in your development configuration, and you'd need to set the setting mentioned earlier (`SENTRY_WEBHOOK_LOG_REDIS_CLUSTER`) to use it.

Reply:
> I guess my comment didn't go here like I had intended, so please refer to this: [#15269 (comment)](https://github.com/getsentry/sentry/pull/15269#issuecomment-548077147)

Is the implication here that `redis.redis_clusters` needs to also handle the case where it's a single node non-Redis Cluster, and return a `StrictRedis` instance, instead of a  `StrictRedisCluster` instance?

Reply:
Yeah. I don't recall off the top of my head if it's possible to have a single node "redis cluster" cluster. In which case, the number of nodes isn't clearly an indicator. i think configuration needs like, `use_cluster=True` to go with it. I think this came up recently with semaphore/relay configuration and we learned you can run a single node redis cluster.

I know, this is confusing. Which is why I try to frame this in terms of the API. Since redis and redis-cluster are a compatible (mostly) API, we just need to handle "redis" as the API. The implementation just which class is spit back. Whether it's cluster or not, the way you use it is the same.

rb is the odd ball in that the API is entirely different.